### ISSUE TITLE
`git diff --no-index `  output paths are not formatted correctly

### ISSUE THREAD

**Issue by rahul-ve** [git-for-windows/git#3357]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.31.1.windows.1
cpu: x86_64
built from commit: c5f0be26a7e3846e3b6268d1c6c4800d838c6bbb
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.19043.1151]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
$ cat "C:\Program Files\Git\etc\install-options.txt"


Editor Option: VisualStudioCode
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Core
Performance Tweaks FSCache: Enabled
Enable Symlinks: Disabled
Enable Pseudo Console Support: Disabled
```

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

This issue is consistent across PowerShellv5, PowerShell core v7, CMD

 - What commands did you run to trigger this issue? 

I am trying to generate a diff of two folders that have similar files. 
There is an issue with `git diff --no-index` output, mainly rendering of paths, so when trying to apply patch it fails as the paths are mangled 
```
# lets take a simple folder structure as below 

├── project1
│   └── folder1
│       └── simple_test.sh
└── project2
    └── folder2
        └── simple_test.sh

# git diff output is as below

$ git diff --no-index --  .\project1\folder1\  .\project2\folder2\

diff --git "a/.\\project1\\folder1\\/simple_test.sh" "b/.\\project2\\folder2\\/simple_test.sh"
index 349b8b3..2f1a918 100644
--- "a/.\\project1\\folder1\\/simple_test.sh"
+++ "b/.\\project2\\folder2\\/simple_test.sh"
@@ -1,3 +1,3 @@
 #!/usr/bin/env bash

-echo "blah blah BLAH!"
+echo "Hello world!"


# and I normally save this output to a patch and apply it to project1 folder, but throws an error

$ cd project1

$ git apply -p2   ..\fix.patch
error: simple_test.sh: No such file or directory

```
 - What did you expect to occur after running these commands?

This process works fine when using Bash (either Git Bash on Windows or Linux), looks like something to do with Windows shells.

Thank you!

**Comment by dscho** [comment]
> $ git diff --no-index --  .\project1\folder1\  .\project2\folder2\

That won't work in Git Bash: the backslash is an escape character there.

**Comment by dscho** [comment]
> $ git apply -p2   ..\fix.patch
> error: simple_test.sh: No such file or directory

And this looks more as if the file `simple_test.sh` was expected to be in the `project1` directory, but it simply wasn't there.

**Comment by rahul-ve** [comment]
> > $ git diff --no-index --  .\project1\folder1\  .\project2\folder1\
> 
> That won't work in Git Bash: the backslash is an escape character there.

Yes the exact commands dont work, but the process with nix compatible path woks just fine. 
I made a slight mistake in my original post, only the top level folders names are different not the subfolders. I fixed it now, hope it is clear. 

I think it is to do with how the paths are rendered, note the output  once I remove `-p2`
```
$ git apply ..\fix.patch
error: .\project1\folder1\/simple_test.sh: No such file or directory
```

**Comment by dscho** [comment]
Okay, there are probably two issues at play, and I am unsure whether I would consider any of them a bug.

The first thing is that the generated diff contains the awkward sequence `\\/` of _two_ consecutive directory delimiters. Git tries to avoid this:

- https://github.com/git/git/blob/225bc32a989d7a22fa6addafd4ce7dcd04675dbf/diff-no-index.c#L138
- https://github.com/git/git/blob/225bc32a989d7a22fa6addafd4ce7dcd04675dbf/diff-no-index.c#L144

As you can see, it will only append a `/` if there isn't one yet. But on Windows, there might be a backslash, and Git does not care at this point whether that is a directory separator and appends a forward slash anyway.

This _could_ be considered a bug, but it is contestable because the _actual_ root cause of the issue at hand is that Git really considers forward slashes as directory separators in its diff machinery. Backslashes as potential directory separators are only an afterthought in Git and not well supported.

The second issue is that not only the diff machinery considers forward slashes as directory separators, but also the apply machinery. See for yourself:

- https://github.com/git/git/blob/225bc32a989d7a22fa6addafd4ce7dcd04675dbf/apply.c#L1119
- https://github.com/git/git/blob/225bc32a989d7a22fa6addafd4ce7dcd04675dbf/apply.c#L1124

These two lines are in the code that handles `-p2`. As you can see, it does not even consider the backslashes as directory separators. As a consequence, too much is stripped of your path.

So what is the correct resolution here?

I guess we won't be able to convince the Git project that backslashes should be handled as if they were directory separators in diffs, because that would break cross-platform compatibility.

The best way forward, therefore, is probably to convert the backslashes to forward slashes on Windows. Something like this:

```diff
diff --git a/diff-no-index.c b/diff-no-index.c
index 9a8b09346bd..c19091366f1 100644
--- a/diff-no-index.c
+++ b/diff-no-index.c
@@ -244,6 +244,7 @@ int diff_no_index(struct rev_info *revs,
 {
 	int i, no_index;
 	const char *paths[2];
+	char *prefixed[2] = { NULL, NULL };
 	struct strbuf replacement = STRBUF_INIT;
 	const char *prefix = revs->prefix;
 	struct option no_index_options[] = {
@@ -272,8 +273,11 @@ int diff_no_index(struct rev_info *revs,
 			 * path that is "-", spell it as "./-".
 			 */
 			p = file_from_standard_input;
-		else if (prefix)
-			p = prefix_filename(prefix, p);
+		else
+			p = prefixed[i] = !prefix ?
+				xstrdup(p) : prefix_filename(prefix, p);
+		if (prefixed[i])
+			convert_slashes(prefixed[i]);
 		paths[i] = p;
 	}
 
@@ -301,6 +305,8 @@ int diff_no_index(struct rev_info *revs,
 	diff_flush(&revs->diffopt);
 
 	strbuf_release(&replacement);
+	FREE_AND_NULL(prefixed[0]);
+	FREE_AND_NULL(prefixed[1]);
 
 	/*
 	 * The return code for --no-index imitates diff(1):
```

@rahul-ve how about making this your first contribution to this project? All you need to do is to provide a good commit message and a test case (you already came up with it, now you only have to turn it into a small shell script snippet, say, in `t/t4053-diff-no-index.sh`, using the `MINGW` prerequisite, [this wiki page](https://github.com/git-for-windows/git/wiki/Running-Git's-regression-tests#running-individual-tests) should help develop this). Oh, and then open a PR ;-)

**Comment by rahul-ve** [comment]
Thanks @dscho for looking to this issue.
> how about making this your first contribution to this project?
Not a problem, would love to.
